File: hosts/mock-ts/src/transport/wire-envelope.ts
// quackoscope-host-mock -- transport layer.
//
// JSON envelope encode/decode and the 17-byte binary data-frame header, byte
// for byte the same shapes hosts/cpp/src/transport/wire.cpp writes. Nothing
// here knows what a device is.

export interface WireRequest {
  id: number;
  method: string;
  params: Record<string, unknown>;
}

export interface RequestOutcome {
  ok: boolean;
  result?: unknown;
  code?: string;
  detail?: string;
}

export function encodeResult(id: number, result: unknown): string {
  return JSON.stringify({ id, result: result === undefined ? null : result });
}

export function encodeError(id: number, code: string, detail: string): string {
  return JSON.stringify({ id, error: { code, detail } });
}

/**
 * Returns the decoded request, or a complete error envelope ready to send back.
 * A request whose id cannot be read gets id 0, exactly as the C++ host does,
 * because there is nothing else to correlate it with.
 */
export function decodeRequest(text: string): { request?: WireRequest; errorEnvelope?: string } {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (e) {
    return { errorEnvelope: encodeError(0, "invalid_value", `request is not valid JSON: ${String(e)}`) };
  }
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    return { errorEnvelope: encodeError(0, "invalid_value", "request is not a JSON object") };
  }
  const envelope = parsed as Record<string, unknown>;
  const id = typeof envelope.id === "number" ? envelope.id : 0;
  if (typeof envelope.method !== "string") {
    return { errorEnvelope: encodeError(id, "invalid_value", 'request is missing a string "method"') };
  }
  const params =
    typeof envelope.params === "object" && envelope.params !== null && !Array.isArray(envelope.params)
      ? (envelope.params as Record<string, unknown>)
      : {};
  return { request: { id, method: envelope.method, params } };
}

export const DATA_FRAME_HEADER_BYTES = 17;

/**
 * Little-endian, 17 bytes, then the payload doubles:
 *   uint32 subscription_id
 *   uint64 domain_start
 *   uint32 sample_count
 *   uint8  encoding      (0 = raw, 1 = min_max_envelope)
 */
export function encodeDataFrame(
  subscriptionId: number,
  domainStart: bigint,
  sampleCount: number,
  encoding: number,
  payload: Float64Array,
): Buffer {
  const frame = Buffer.allocUnsafe(DATA_FRAME_HEADER_BYTES + payload.length * 8);
  frame.writeUInt32LE(subscriptionId, 0);
  frame.writeBigUInt64LE(domainStart, 4);
  frame.writeUInt32LE(sampleCount, 12);
  frame.writeUInt8(encoding, 16);
  for (let i = 0; i < payload.length; i++) {
    frame.writeDoubleLE(payload[i], DATA_FRAME_HEADER_BYTES + i * 8);
  }
  return frame;
}

File: hosts/csharp/src/Transport/WireEnvelope.cs
// Quackoscope host (C#) -- transport layer.
//
// JSON envelope encode/decode and binary data-frame framing for the M1 wire
// contract. Nothing under Transport/ may reference an openDAQ type.
using System.Buffers.Binary;
using System.Text.Json.Nodes;

namespace Quackoscope.Host.CSharp.Transport;

// What a request handler gives back. The transport turns this into either
// {"id": n, "result": ...} or {"id": n, "error": {"code":..., "detail":...}}.
public sealed class RequestOutcome
{
    public bool Ok = true;
    public JsonNode Result;
    public string Code = "";
    public string Detail = "";
}

public sealed class DecodedRequest
{
    public long Id;
    public string Method = "";
    public JsonObject Params = new();
}

public static class WireEnvelope
{
    public const int DataFrameHeaderBytes = 17;

    public static bool TryDecodeRequest(string text, out DecodedRequest request, out JsonNode errorEnvelope)
    {
        request = null;
        errorEnvelope = null;

        JsonNode parsed;
        try
        {
            parsed = JsonNode.Parse(text);
        }
        catch (Exception e)
        {
            errorEnvelope = EncodeError(0, "invalid_value", $"message is not JSON: {e.Message}");
            return false;
        }

        if (parsed is not JsonObject envelope)
        {
            errorEnvelope = EncodeError(0, "invalid_value", "message is not a JSON object");
            return false;
        }

        long id = 0;
        if (envelope.TryGetPropertyValue("id", out var idNode) && idNode is JsonValue idValue && idValue.TryGetValue(out long parsedId))
            id = parsedId;
        else
        {
            errorEnvelope = EncodeError(0, "invalid_value", "envelope has no integer \"id\"");
            return false;
        }

        if (!envelope.TryGetPropertyValue("method", out var methodNode) ||
            methodNode is not JsonValue methodValue ||
            !methodValue.TryGetValue(out string method))
        {
            errorEnvelope = EncodeError(id, "invalid_value", "envelope has no string \"method\"");
            return false;
        }

        var parameters = new JsonObject();
        if (envelope.TryGetPropertyValue("params", out var paramsNode) && paramsNode is JsonObject given)
            parameters = (JsonObject)given.DeepClone();

        request = new DecodedRequest { Id = id, Method = method, Params = parameters };
        return true;
    }

    public static JsonNode EncodeResult(long id, JsonNode result) =>
        new JsonObject { ["id"] = id, ["result"] = result?.DeepClone() };

    public static JsonNode EncodeError(long id, string code, string detail) =>
        new JsonObject
        {
            ["id"] = id,
            ["error"] = new JsonObject { ["code"] = code, ["detail"] = detail }
        };

    // 17-byte little-endian header, then the payload doubles:
    //   uint32 subscription_id
    //   uint64 domain_start
    //   uint32 sample_count
    //   uint8  encoding      (0 = raw, 1 = min_max_envelope)
    public static byte[] EncodeDataFrame(uint subscriptionId,
                                         ulong domainStart,
                                         uint sampleCount,
                                         byte encoding,
                                         ReadOnlySpan<double> payload)
    {
        var frame = new byte[DataFrameHeaderBytes + payload.Length * sizeof(double)];
        var span = frame.AsSpan();
        BinaryPrimitives.WriteUInt32LittleEndian(span[..4], subscriptionId);
        BinaryPrimitives.WriteUInt64LittleEndian(span.Slice(4, 8), domainStart);
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(12, 4), sampleCount);
        span[16] = encoding;
        for (var i = 0; i < payload.Length; i++)
            BinaryPrimitives.WriteDoubleLittleEndian(span.Slice(DataFrameHeaderBytes + i * sizeof(double), sizeof(double)), payload[i]);
        return frame;
    }
}
